[two_dimensional_scrollables] Adds generics to the callbacks and builders of TreeView

baumths opened the pull request:
This PR adds generics to the callbacks and builders of the TreeView widget. This removes the need to type cast the content of a TreeViewNode.

Fixes https://github.com/flutter/flutter/issues/149656

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

Piinks reviewed (commented):
Hey @baumths thanks for sending a PR! You beat me to it! :)

Can you add a test that prevents us from regressing this in the future? Just creating an instance of those callbacks and builders that does not have the type casts should cover it, like in the comment you left on the TreeSliver PR.

baumths commented:
Added a test but I'm not sure if I did it the right way, never written such a test. Should I expect something else?

I've also updated the tree view tests to use the correct type in the callbacks, hope that's ok.

Piinks commented on packages/two_dimensional_scrollables/pubspec.yaml:
```diff
@@ -1,6 +1,6 @@
 name: two_dimensional_scrollables
 description: Widgets that scroll using the two dimensional scrolling foundation.
-version: 0.3.0
+version: 0.3.0+1
```
```suggestion
version: 0.3.1
```

Worthy of the bump!

Piinks reviewed (commented):
Test coverage is great thank you! Can you give it a minor version bump instead of +1?

Piinks reviewed (approved):
![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/169392212-02bb74fe-c6b9-4eb7-aa62-6ddb43a70d5b.png)

Piinks commented:
I need to get a second stamp for this, will find another reviewer

matanlurey reviewed (approved):
LGTM.

I had a question about whether adding generics is considered a breaking change, but it was answered ("no"). Thanks!
